### PR TITLE
test: add preliminary functional test coverage

### PR DESCRIPTION
**Pull request by jon-whit**
## Description
This PR starts laying the foundation for functional test coverage of the OpenFGA grpc and HTTP servers. The changes herein work by launching an `openfga/openfga` container (built ahead of time) and running tests against the grpc and HTTP servers by interacting with the APIs via a client connected to the running container.

In the CI/CD pipeline we build the openfga image based on the changes included in the PR and we run the functional test suite against it.

## Review Checklist
- [x] I have added documentation for new/changed functionality in this PR or in [openfga.dev](https://github.com/openfga/openfga.dev)
- [X] The correct base branch is being used, if not `main`

### CODE REVIEW

File: cmd/openfga/openfga_test.go
Comment (jon-whit) [inline cmd/openfga/openfga_test.go:154]: This probably belongs in our `testutils` package instead of here.

- [ ] todo(jon-whit): move this to the `testutils` package.

File: cmd/openfga/openfga_test.go
Comment (jon-whit) [inline cmd/openfga/openfga_test.go:137]: - [ ] todo(jon-whit): wire up a health check here.

File: Makefile
Comment (jon-whit) [inline Makefile:52]: Not sure the best way to approach this. Ideally you could just run `make functional-test` and it would 1) build the docker image against the latest changes and 2) run the functional tests (by passing the `-tags=functional` flag).

If we naively do

```
.PHONY: build-image
build-image: go-generate
	docker build -t openfga/openfga:latest .

.PHONY: functional-test
functional-test: build-image
	go test -v ...
```
then every time we invoke it we will have to go through the docker build process, which doesn't cache the `go build ...` step. This is pretty time costly, and in a CI/CD environment we pre-build the image anyways.

I'm wondering if it's ok to just assume the image is pre-built and if you want to run functional tests locally you'll have to make sure you do so. What do you think 🤷 ?

File: cmd/openfga/openfga_test.go
Comment (jon-whit) [inline cmd/openfga/openfga_test.go:2]: This is a Go build tag, which means that this file will only get included in the test build if you pass the `-tags=functional` flag to `go test`. The idea I was going for is that, by default, functional tests don't run. You have to specify the build tag to get the functional tests to run.

File: cmd/openfga/openfga_test.go
Comment (jon-whit) [inline cmd/openfga/openfga_test.go:200]: These tests can kind of be broken down into 2 categories: tests for input validation and tests for standard workflows (functional workflows). The former is a great candidate for test tables, while the later becomes a little more challenging for table driven tests because of the preconditions that have to be met in many cases.

What do you think about creating two separate tests for these different categories? Any preference on this structure?

File: Makefile
Comment (craigpastro) [reply Makefile:52]: Both of these choices are not ideal. Do you think it is not enough to run it against `main`?

File: cmd/openfga/openfga_test.go
Comment (craigpastro) [inline cmd/openfga/openfga_test.go:42]: Do we need this function? Or can we just make `grpcPort` in `serverHandle` public?

File: cmd/openfga/openfga_test.go
Comment (craigpastro) [inline cmd/openfga/openfga_test.go:92]: Should we create a single ctx instead of using `context.Background()` everywhere?

File: cmd/openfga/openfga_test.go
Comment (craigpastro) [inline cmd/openfga/openfga_test.go:139]: Shall we remove this?

File: cmd/openfga/openfga_test.go
Comment (craigpastro) [inline cmd/openfga/openfga_test.go:163]: Do we need this? Or should we just set `MaxElapsedTime` to 2 seconds above?

File: cmd/openfga/openfga_test.go
Comment (jon-whit) [reply cmd/openfga/openfga_test.go:93]: Sure!

File: cmd/openfga/openfga_test.go
Comment (jon-whit) [reply cmd/openfga/openfga_test.go:163]: I don't want the `Check` to hang until `MaxElapsedTime` has been reached. I want to limit how long the Check can take and then retry the check again in a subsequent retry. If we didn't set the timeout ctx then the single Check could hang until `MaxElapsedTime` elapsed.

File: cmd/openfga/openfga_test.go
Comment (craigpastro) [reply cmd/openfga/openfga_test.go:43]: Sounds good to me. Thanks!

File: Makefile
Comment (jon-whit) [reply Makefile:52]: So I originally added the `build-image` Makefile target and had it as a pre-req for this target, but I really didn't like that because every time you run `make functional-test` it does a `docker build ...` which has all of the docker build output, and it takes longer than it probably needs.

The problem is that the `build-image` target shouldn't probably be run if there aren't any sources that have changed that would affect the built image. So to improve that experience we'd want to only have the `build-image` target run if any of the Go source files have changed or any of the artifacts included in the container.

We could do that, it just requires a good chunk of more work to orchestrate that through the Makefile (but it is what make was made for, haha).

WDYT, should we bite the bullet and do ^^?

**Comment by jon-whit** [comment]
Tests will fail until #105 is fixed. I've added tests to assert the correct behavior here. I'd like #105 to be merged before merging this.

File: Makefile
Comment (miparnisari) [reply Makefile:52]: > The problem is that the `build-image` target shouldn't probably be run if there aren't any sources that have changed that would affect the built image.

For local development, I will always have to build the image, no? Because presumably I changed something and I want to see if the tests still pass.

File: cmd/openfga/openfga_test.go
Comment (miparnisari) [inline cmd/openfga/openfga_test.go:115]: is this the time we expect the tests to take? if so, i think it's a little low... 10 minutes i've seen them take

File: cmd/openfga/openfga_test.go
Comment (miparnisari) [inline cmd/openfga/openfga_test.go:200]: i'm confused as to why we have `GRPCCreateStoreTest` and `HTTPCreateStoreTest` - are we going to duplicate the code? 🤯

Comment (jon-whit) [reply cmd/openfga/openfga_test.go:200]: They are both part of the API surface, and given that we use middleware that differs between the two I think it's important we test the functional API behavior of both :).

So yeah, we'll test both.

File: Makefile
Comment (jon-whit) [reply Makefile:52]: @miparnisari 

> For local development, I will always have to build the image, no? Because presumably I changed something and I want to see if the tests still pass.

Most of the time, yes. But not always. If none of the Go files or artifacts that go inside the container have changed, then you don't need to rebuild the image to run functional tests.

> Can you remind me why we did it like this? I forgot :/

Because this gives us the highest fidelity functional test coverage. This approach allows us to test the entrypoint(s) of the container as well as the server code embedded in it. If we only tested the server implementation without spinning up the container, then we could easily foobar the Dockerfile and introduce something unintended. This approach allows us to test the whole integration end to end with very little overhead (you just have to build the container).

File: cmd/openfga/openfga_test.go
Comment (jon-whit) [reply cmd/openfga/openfga_test.go:115]: If a test hasn't completed in 2min then we have a problem. Tests should have stricter timeouts by using `context.WithTimeout(...)`.

File: cmd/openfga/openfga_test.go
Comment (jon-whit) [inline cmd/openfga/openfga_test.go:240]: This makes no sense to me as a client and as a maintainer of this project, haha 🤣 

What I should be asserting here is the following:

```
require.Equal(t, codes.Unauthenticated, s.Code())
```
We should adhere to the canonical status codes of the protocol we are using. It makes no sense for us to be implementing our own status code protocol ontop of a preexisting one.

File: server/authn/oidc/oidc.go
Comment (elbuo8) [inline server/authn/oidc/oidc.go:14]: Why this change?

File: cmd/openfga/openfga_test.go
Comment (elbuo8) [inline cmd/openfga/openfga_test.go:961]: I believe a util was recently introduced to this.

File: .github/workflows/pull_request.yaml
Comment (elbuo8) [inline .github/workflows/pull_request.yaml:88]: Could you please document this so that it doesn't get randomly flipped?

File: .github/workflows/pull_request.yaml
Comment (elbuo8) [inline .github/workflows/pull_request.yaml:89]: Perhaps are more meaningful tag?

File: cmd/openfga/openfga_test.go
Comment (jon-whit) [reply cmd/openfga/openfga_test.go:961]: Oops. That's residual. It's in the `tuple` package.

File: .github/workflows/pull_request.yaml
Comment (jon-whit) [reply .github/workflows/pull_request.yaml]: I switched it to `openfga/openfga:functionaltest`.

File: .github/workflows/pull_request.yaml
Comment (jon-whit) [reply .github/workflows/pull_request.yaml]: Added a comment to it

File: server/authn/oidc/oidc.go
Comment (elbuo8) [reply server/authn/oidc/oidc.go:14]: For a subsequent PR can we add a linting rule to prevent this from happening and normalize?

File: .github/workflows/pull_request.yaml
Comment (elbuo8) [inline .github/workflows/pull_request.yaml:67]: This isn't needed with the other change, remove in a subsequent PR to get this one in @jon-whit

**Review by elbuo8 — APPROVED** [review]

**Pull request merged**
